### PR TITLE
Improve join order for RangeAnalysis::boundedPhi

### PR DESCRIPTION
**Pull request by smowton**

### CODE REVIEW

File: java/ql/lib/semmle/code/java/dataflow/RangeAnalysis.qll [hunk 1]
```diff
@@ -734,7 +744,11 @@ private predicate boundedPhi(
   SsaPhiNode phi, Bound b, int delta, boolean upper, boolean fromBackEdge, int origdelta,
   Reason reason
 ) {
-  forex(SsaVariable inp, SsaReadPositionPhiInputEdge edge | edge.phiInput(phi, inp) |
+  // Forcibly split into an existence and a forall stage instead of using forex because otherwise
```
Comment (hvitved) [inline java/ql/lib/semmle/code/java/dataflow/RangeAnalysis.qll:747]: I am surprised to see that this doesn't use the "ranking approach" as in e.g. the [modulus library](https://github.com/github/codeql/blob/main/java/ql/lib/semmle/code/java/dataflow/ModulusAnalysis.qll#L162-L184). Replacing recursion through `forall`/`forex` with recursion over ranked indices typically performs much better.

Comment (smowton) [reply java/ql/lib/semmle/code/java/dataflow/RangeAnalysis.qll:747]: Like https://github.com/github/codeql/pull/8405 ? Which do you prefer?

Comment (hvitved) [reply java/ql/lib/semmle/code/java/dataflow/RangeAnalysis.qll:747]: Yes. I prefer that PR over this one.

Comment (hvitved) [reply java/ql/lib/semmle/code/java/dataflow/RangeAnalysis.qll:747]: (this one also appears to have a significant slow-down on `palatable/lambda`).

**Comment by smowton** [comment]
Dropped in favour of #8405

**Pull request closed without merging**